File: src/main/java/org/tobi29/scapes/connection/Account.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.connection;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tobi29.scapes.engine.utils.RSAUtil;
import org.tobi29.scapes.engine.utils.UnsupportedJVMException;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.NoSuchAlgorithmException;
import java.security.PublicKey;
import java.security.interfaces.RSAPrivateCrtKey;
import java.security.spec.InvalidKeySpecException;
import java.util.Optional;
import java.util.Properties;

public class Account {
    private static final Logger LOGGER = LoggerFactory.getLogger(Account.class);

    public static Client read(Path path) throws IOException {
        String key = null, nickname = "";
        if (Files.exists(path)) {
            Properties properties = new Properties();
            try (InputStream streamIn = Files.newInputStream(path)) {
                properties.load(streamIn);
            }
            key = properties.getProperty("Key");
            nickname = properties.getProperty("Nickname", "");
        }
        Optional<KeyPair> keyPair = key(key);
        if (keyPair.isPresent()) {
            return new Client(keyPair.get(), nickname);
        }
        return new Client(genKey(), nickname);
    }

    public static Optional<KeyPair> key(String str) {
        if (str == null) {
            return Optional.empty();
        }
        try {
            RSAPrivateCrtKey privateKey = RSAUtil.readPrivate(str);
            PublicKey publicKey = RSAUtil.extractPublic(privateKey);
            return Optional.of(new KeyPair(publicKey, privateKey));
        } catch (InvalidKeySpecException | IllegalArgumentException e) {
            LOGGER.warn("Failed to parse key: {}", e.toString());
            return Optional.empty();
        }
    }

    public static KeyPair genKey() {
        LOGGER.info("Generating key-pair...");
        try {
            KeyPairGenerator keyPairGenerator =
                    KeyPairGenerator.getInstance("RSA");
            keyPairGenerator.initialize(4096);
            return keyPairGenerator.genKeyPair();
        } catch (NoSuchAlgorithmException e) {
            throw new UnsupportedJVMException(e);
        }
    }

    public static String key(KeyPair keyPair) {
        if (keyPair == null) {
            return "";
        }
        try {
            return RSAUtil.writePrivate(keyPair.getPrivate());
        } catch (InvalidKeySpecException e) {
            return "";
        }
    }

    public static boolean valid(String nickname) {
        return !isNameValid(nickname).isPresent();
    }

    public static Optional<String> isNameValid(String nickname) {
        if (nickname.length() < 6) {
            return Optional.of("Name must be at least 6 characters long!");
        }
        if (nickname.length() > 20) {
            return Optional.of("Name may not be longer than 20 characters!");
        }
        if (nickname.contains(" ")) {
            return Optional.of("Name may not contain spaces!");
        }
        for (int i = 0; i < nickname.length(); i++) {
            if (!Character.isLetterOrDigit(nickname.charAt(i))) {
                return Optional.of("Name may only contain letters and digits!");
            }
        }
        return Optional.empty();
    }

    public static class Client {
        private final KeyPair keyPair;
        private final String nickname;

        public Client(KeyPair keyPair, String nickname) {
            this.keyPair = keyPair;
            this.nickname = nickname;
        }

        public void write(Path path) throws IOException {
            Properties properties = new Properties();
            properties.setProperty("Key", key(keyPair));
            properties.setProperty("Nickname", nickname);
            try (OutputStream streamOut = Files.newOutputStream(path)) {
                properties.store(streamOut, "");
            }
        }

        public KeyPair getKeyPair() {
            return keyPair;
        }

        public String getNickname() {
            return nickname;
        }

        public boolean valid() {
            return Account.valid(nickname);
        }
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/connection/Account.kt
/*
 * Copyright 2012-2016 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.connection

import mu.KLogging
import org.tobi29.scapes.engine.utils.UnsupportedJVMException
import org.tobi29.scapes.engine.utils.extractPublic
import org.tobi29.scapes.engine.utils.io.ByteStreamInputStream
import org.tobi29.scapes.engine.utils.io.ByteStreamOutputStream
import org.tobi29.scapes.engine.utils.io.filesystem.FilePath
import org.tobi29.scapes.engine.utils.io.filesystem.exists
import org.tobi29.scapes.engine.utils.io.filesystem.read
import org.tobi29.scapes.engine.utils.io.filesystem.write
import org.tobi29.scapes.engine.utils.readPrivate
import org.tobi29.scapes.engine.utils.writePrivate
import java.io.IOException
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.NoSuchAlgorithmException
import java.security.spec.InvalidKeySpecException
import java.util.*

class Account(private val keyPair: KeyPair, private val nickname: String) {

    @Throws(IOException::class)
    fun write(path: FilePath) {
        val properties = Properties()
        properties.setProperty("Key", key(keyPair))
        properties.setProperty("Nickname", nickname)
        write(path) {
            properties.store(ByteStreamOutputStream(it), "")
        }
    }

    fun keyPair(): KeyPair {
        return keyPair
    }

    fun nickname(): String {
        return nickname
    }

    fun valid(): Boolean {
        return valid(nickname)
    }

    companion object : KLogging() {

        @Throws(IOException::class)
        operator fun get(path: FilePath): Account {
            val account = read(path) ?: throw IOException(
                    "No valid account in: " + path)
            return account
        }

        @Throws(IOException::class)
        fun read(path: FilePath): Account? {
            var key: String? = null
            var nickname = ""
            if (exists(path)) {
                val properties = Properties()
                read(path, {
                    properties.load(ByteStreamInputStream(it))
                })
                key = properties.getProperty("Key")
                nickname = properties.getProperty("Nickname", "")
            }
            val keyPair = key(key)
            if (keyPair != null) {
                return Account(keyPair, nickname)
            }
            return null
        }

        @Throws(IOException::class)
        fun generate(path: FilePath): Account {
            val account = Account(genKey(), "")
            account.write(path)
            return account
        }

        fun key(str: String?): KeyPair? {
            if (str == null) {
                return null
            }
            try {
                val privateKey = readPrivate(str)
                val publicKey = extractPublic(privateKey)
                return KeyPair(publicKey, privateKey)
            } catch (e: InvalidKeySpecException) {
                logger.warn { "Failed to parse key: $e" }
                return null
            } catch (e: IllegalArgumentException) {
                logger.warn { "Failed to parse key: $e" }
                return null
            }

        }

        fun genKey(): KeyPair {
            logger.info { "Generating key-pair..." }
            try {
                val keyPairGenerator = KeyPairGenerator.getInstance("RSA")
                keyPairGenerator.initialize(4096)
                return keyPairGenerator.genKeyPair()
            } catch (e: NoSuchAlgorithmException) {
                throw UnsupportedJVMException(e)
            }

        }

        fun key(keyPair: KeyPair?): String {
            if (keyPair == null) {
                return ""
            }
            try {
                return writePrivate(keyPair.private)
            } catch (e: InvalidKeySpecException) {
                return ""
            }

        }

        fun valid(nickname: String): Boolean {
            return isNameValid(nickname) == null
        }

        fun isNameValid(nickname: String): String? {
            if (nickname.length < 6) {
                return "Name must be at least 6 characters long!"
            }
            if (nickname.length > 20) {
                return "Name may not be longer than 20 characters!"
            }
            if (nickname.contains(" ")) {
                return "Name may not contain spaces!"
            }
            for (i in 0..nickname.length - 1) {
                if (!Character.isLetterOrDigit(nickname[i])) {
                    return "Name may only contain letters and digits!"
                }
            }
            return null
        }
    }
}
